[Foxy] Move Aggregator publishing to timer to allow subscription callback more processing time

cdbierl opened the pull request:
Moving the aggregator's publishing to a wall timer, and letting the node spin continuously, permits the subscription callback to operate at the necessary rate of incoming diagnostic entry messages.

cdbierl committed:
Merge pull request #1 from ros/foxy

Foxy

Commit:
refactored main method and moved publishData to wall timer (to fix subscription callback falling behind)

Karsten1987 commented:
This generally looks good to me. However, please be advised that this branch has to target `ros2-devel` and not `foxy`. We can't accept this patch in `foxy` as it's breaking API.

Also @norro you might want to weigh in here as the author of the aggregators ;-)

norro commented on diagnostic_aggregator/src/aggregator.cpp:
```diff
@@ -101,6 +102,10 @@ Aggregator::Aggregator()
     n_->create_publisher<DiagnosticStatus>("/diagnostics_toplevel_state", 1);
+
+  int publish_rate_ms = 1000 / pub_rate_;
+  publish_timer_ = n_->create_wall_timer(std::chrono::milliseconds(publish_rate_ms),
```
please uncrustify:
```
-  publish_timer_ = n_->create_wall_timer(std::chrono::milliseconds(publish_rate_ms),
+  publish_timer_ = n_->create_wall_timer(
+    std::chrono::milliseconds(publish_rate_ms),

norro commented on diagnostic_aggregator/src/aggregator_node.cpp:
```diff
@@ -34,34 +34,20 @@
 /**< \author Kevin Watts */
 
-#include <exception>
+#include <rclcpp/rclcpp.hpp>
```
Please include <memory> for using `make_shared`

cdbierl replied:
Are you referring to the original try-catch?

norro replied:
Sorry, the github syntax swallowed parts of the message.
I was asking to `#include <memory>` because you introduce `make_shared`.

norro reviewed (changes requested):
Thank you very much for this PR, it looks good to me as well!
Can you please address the two minor issues detected by cpplint resp. uncrustify and redirect this PR to ros2-devel?

cdbierl commented:
Any objections to me also bringing in the subscription history depth change ```.keep_last(1000)``` in the ROS2 PR ?

Commit:
uncrustify and include memory header

Commit:
cpplint: fix header order

cdbierl commented:
> This generally looks good to me. However, please be advised that this branch has to target `ros2-devel` and not `foxy`. We can't accept this patch in `foxy` as it's breaking API.
> 
> Also @norro you might want to weigh in here as the author of the aggregators ;-)

What is the breaking API change you mention for Foxy?

norro commented:
@cdbierl `publishData()` seems to be private now after your PR.

cdbierl commented:
@norro That is the breaking API change? I also removed some other public functions no longer used in this package. Is the point that other people using the aggregator will find the breaking API change(s)? If you mean I have introduced local breaking API changes, then I understand. I can revise. I am not sure though if this is what @Karsten1987 meant.

Maybe I am coming around to the idea now. In Foxy, you aren't allowed to introduce breaking API changes, like I did by changing access level or removing API functions of the Aggregator class (annotated by the macros)? I suppose this is also the case for my ros2-devel PR? Let me know if I am on the right page now.

Commit:
revert publishData API change

Karsten1987 commented on diagnostic_aggregator/src/aggregator.cpp:
```diff
@@ -42,6 +42,7 @@
 #include <mutex>
 #include <string>
 #include <vector>
+#include <chrono>
```
please sort the includes alphabetically.

Karsten1987 commented on diagnostic_aggregator/src/aggregator_node.cpp:
```diff
@@ -34,34 +34,21 @@
-
-using std::exception;
+#include <memory>
+#include <rclcpp/rclcpp.hpp>
```
```suggestion

#include "rclcpp/rclcpp.hpp"
```

cdbierl replied:
Would you happen to know if there is a .clang-format file that follows your pipeline checks?

norro replied:
I think you find it in the `ament_clang_format` package resp. in your ROS 2 workspace in `src/ament/ament_lint/ament_clang_format/ament_clang_format/configuration/.clang-format`

Karsten1987 reviewed (approved):
Looks good to me with two nitpicks addressed and green CI (uncrustify, cppcheck)

Karsten1987 commented:
> What is the breaking API change you mention for Foxy?

You have removed functions and moved some as private. That introduces a breaking API change where external users can't rely on these functions any longer within the same ros-distro.
It actually goes one step further where we don't want to even have ABI breaks, so that external users don't have to recompile their libraries using diagnostics in Foxy.
That is to say, new changes should always land on ros2-devel so that we don't have to care much about API/ABI breakages. If you want to use binary packages (debians) for the latest state of ROS2, we recommend to rely on the `rolling` distro.

> Any objections to me also bringing in the subscription history depth change .keep_last(1000) in the ROS2 PR ?

Let's keep the changes semantically separated. There's already a PR open for this and I'd rather push forward on that one than further diluting this PR.

norro reviewed (approved):
LGTM

Commit:
fixes for header include conventions

cdbierl commented:
Is there anything I need to do from the Rpr__diagnostics__ubuntu_focal_amd64 test failure? Is that trying to build this for Rolling?

Karsten1987 commented:
~so I've changed the target branch to `ros2-devel` as discussed above. Do you might taking a look at the conflicts? I am confident that a simple `git rebase` should do the trick.~

My bad: Just realized that this PR deliberately targets `Foxy`. Please be advised then that the current status can't be merged due to API/ABI changes.

Karsten1987 reviewed (changes requested):
can't be merged due to API/ABI changes.

cdbierl commented:
@Karsten1987 I'm not sure I understand what needs to be adjusted. The only thing I see failing is this not building for Rolling in CI, but that has nothing to do with this changeset. What particularly is not valid?

Karsten1987 commented:
> You have removed functions and moved some as private. That introduces a breaking API change where external users can't rely on these functions any longer within the same ros-distro.
It actually goes one step further where we don't want to even have ABI breaks, so that external users don't have to recompile their libraries using diagnostics in Foxy.

So you can't remove the two functions (API break) and you can't really add the timer base shared pointer to the header (ABI break). 
The KDE community has a quite helpful [website](https://community.kde.org/Policies/Binary_Compatibility_Issues_With_C%2B%2B) about compatibility.

Commit:
restored aggregator header, moved timer creation to main for scope workaround

Karsten1987 reviewed (approved):
that looks better now. Thanks for iterating with me over this.
